ReportUtilityBills/measure.rb

```diff
@@ -0,0 +1,785 @@
+    rows = rows[2..-1]
+    headers = rows[0]
+
+    rows = rows.reverse[1..26].transpose
+    rows.each_with_index do |row, i|
+      marginal_rates[headers[i]] = row
+    end
+
+    marginal_rates.delete('Date')
+    marginal_rates.each do |header, values|
+      if values.all? { |x| x.nil? }
+        marginal_rates.delete(header)
+        next
+      end
+      marginal_rates[header] = marginal_rates[header].map { |x| Float(x) }
+    end
+
+    return marginal_rates
+  end
+
+  def write_output(runner, utility_bills, output_format, output_path)
+    line_break = nil
+
+    segment, _ = utility_bills.keys[0].split(':', 2)
+    segment = segment.strip
+    results_out = []
+    utility_bills.each do |key, bill|
+      new_segment, _ = key.split(':', 2)
+      new_segment = new_segment.strip
+      if new_segment != segment
+        results_out << [line_break]
+        segment = new_segment
+      end
+
+      results_out << ["#{key}: Fixed ($)", bill.annual_fixed_charge.round(2)] if [FT::Elec, FT::Gas].include? key
+      results_out << ["#{key}: Marginal ($)", bill.annual_energy_charge.round(2)] if [FT::Elec, FT::Gas].include? key
+      results_out << ["#{key}: PV Credit ($)", bill.annual_production_credit.round(2)] if [FT::Elec].include? key
+      results_out << ["#{key}: Total ($)", bill.annual_total.round(2)]
+    end
+
+    if output_format == 'csv'
+      CSV.open(output_path, 'wb') { |csv| results_out.to_a.each { |elem| csv << elem } }
+    elsif output_format == 'json'
+      h = {}
+      results_out.each do |out|
+        next if out == [line_break]
+
+        grp, name = out[0].split(':', 2)
+        h[grp] = {} if h[grp].nil?
+        h[grp][name.strip] = out[1]
+      end
+
+      require 'json'
+      File.open(output_path, 'w') { |json| json.write(JSON.pretty_generate(h)) }
+    end
+    runner.registerInfo("Wrote bills output to #{output_path}.")
+  end
+
+  def teardown
+    @sqlFile.close()
+
+    # Ensure sql file is immediately freed; otherwise we can get
+    # errors on Windows when trying to delete this file.
+    GC.start()
+  end
+
+  def get_timestamps()
+    if timeseries_frequency == 'hourly'
+      interval_type = 1
+    elsif timeseries_frequency == 'daily'
+      interval_type = 2
+    elsif timeseries_frequency == 'monthly'
+      interval_type = 3
+    elsif timeseries_frequency == 'timestep'
+      interval_type = -1
+    end
+
+    query = "SELECT Year || ' ' || Month || ' ' || Day || ' ' || Hour || ' ' || Minute As Timestamp FROM Time WHERE IntervalType='#{interval_type}'"
+    values = @sqlFile.execAndReturnVectorOfString(query)
+    fail "Query error: #{query}" unless values.is_initialized
+
+    timestamps = []
+    values.get.each do |value|
+      year, month, day, hour, minute = value.split(' ')
+      ts = Time.utc(year, month, day, hour, minute)
+
+      ts_iso8601 = ts.iso8601
+      ts_iso8601 = ts_iso8601.delete('Z')
+      timestamps << ts_iso8601
+    end
+
+    return timestamps
+  end
```

Comment:
Lots of duplicate code with the ReportSimulationOutput measure. I suggest creating a new HPXMLtoOpenStudio/resources/output.rb where you consolidate code and reference from multiple reporting measures. You could also move the existing ReportSimulationOutput/constants.rb code into it and delete that constants.rb file.

Reply:
I started down this route (create `output.rb`, move `constants.rb` into it, and start consolidating code). Can you help me fine tune this file?